Add standard reports docs (#4954)
msupply-foundation/open-msupply: server/report_builder/README.md

By default,

Where the letters are short hand for the following:

- t for translate
  The name of the function
- k for key
  This is the locale key as is used in front end translations.
- n for namespace

>   The file namespace where the translation key is. The .json exention is automattically added ie catalogue (which refers to the catalogue.json namespace).

Reviewer: ```suggestion
  The file namespace where the translation key is. The .json extention is automatically added ie catalogue (which refers to the catalogue.json namespace).
```
But maybe we can remove this line now that we just have `common.json`

Author: Have updated to to specify that we aren't using namespace anymore, but functionality is retained (in case we use customer specific namespaces in future)

  // optional
  // Sub context for reports, applicable to Dispensary and Report context, see client/packages/reports/src/ListView/ListView.tsx",
  // Currently only reports of subcontext 'Expiring' and ' StockAndItems' are displayed
  "sub_context": "StockAndItems",
  // Display name of report
  "name": "Item Usage",
  // optional
  "queries": {
    // optional
    // graphql file query name with extension
    "gql": "query.graphql",
    // optional
    // vec of sql query file names without extension. report_builder will parse both postgres.sql, sqlite.sql, and agnostic .sql file type extensions.
    "sql": ["thisMonthConsumption", "lastMonthConsumption"]
  },
  // optional
  "arguments": {
    // optional
    // location of schema file json relative to the version dir
    "schema": "argument_schemas/arguments.json",
    // optional
    // location of ui schema file json relative to the version dir
    "ui": "argument_schemas/arguments_ui.json"
  }
}
```

This manifest.json file takes the place of cli param inserts seen in bash scripts such as upsert.sh in the open-mSupply-reports repo.

Standard reports in the reports dir can be built into the generated json by running

```bash

> ./target/debug/remote_server_cli build-standard-reports

Reviewer: Are they just built or also upserted? Isn't there two commands you added?
Should we also mention how to build the cli?

```

from the open-msupply/server dir.

## Standard reports versioning

> Standard reports include versions for updates of reports. open mSupply central will automatically sync the most recent standard report compatible with the remote site build of open mSupply.

Reviewer: ```suggestion
Standard reports include versions for updates of reports. open mSupply central will automatically sync all standard report compatible with the remote site build of open mSupply.
```
I think the simplest idea is to sync all reports, then the local version will know which one to use.
It does mean a bit more sync traffic but that is the current plan.

Note: This isn't currently true, so not sure if we should put it in the docs. Maybe it should say from 2.4.0, open-msupply will sync?

> First open mSupply will check for custom reports of the same code which will override any standard report.

Reviewer: Again I don't think it does yet right?

open mSupply central will then sync to the remote site the most recent minor version which matches the major version of the remote site.
For example:

for report_versions = [2.3.0, 2.3.5, 2.8.2, 2.8.3, 3.0.1, 3.5.1]
if remote omSupply.version = 2.3 selected report = 2.3.5
if remote omSupply.version = 2.4 selected report = 2.3.5
if remote omSupply.version = 2.8 selected report = 2.8.3
if remote omSupply.version = 3.2 selected report = 3.0.1
if remote omSupply.version = 4.5 selected report = 3.5.1

### Adding a new standard report version

> Add a new version dir within the name of the report you are adding a new version to.

Reviewer: With the version of open-mSupply that the report is compatible with...

Reviewer: ```suggestion
Add a new version dir within the directory for the report you are adding a new version to.
The version relates to the earliest mSupply version, the report is compatible with.
```

> > Note that the dir names are for developer convenience only. The name and version of each report read by omSupply is from the manifest.json file.

Reviewer: Nice! Good to have these kinds of developer notes, for non-obvious things!
